Feat/f 35 project territories

Voronapxl opened the pull request.

Voronapxl committed:
fix(tasks):
1. context layers are now passed to the before and after methods where called
2. Running tasks are now called correctly
3. Project territories now have flag "is_project:bool"

Voronapxl committed:
feat(effects):
1. F26 is now returning blocks geolayer with properties about services location
2. F35 now returning provision tables
3. 404 exception for when no physical objects are not found on territory
4. Fix for case when provision tables got returned in service_types router

kanootoko commented on app/clients/urban_api_client.py:
```diff
@@ -30,6 +30,10 @@ async def get_physical_objects(
         )
         features = res["features"]
+        if not features:
+            raise http_exception(
+                404, f"Physical objects needed for calculations are not found", params
+            )
```
Вообще, не то, чтобы очень правильно, что клиент к API кидает из себя http_exception. В плане, ты можешь технически использовать клиент и в образном GUI-приложении - и зачем тебе в нем http-ошибки?
И уж тем более "для вычислений" в тексте ошибки. Для каких вычислений, если метод просто получает физические объекты?

То есть, это просто не тот слой, хотя так оно и быстрее пишется (если здесь возвращать по-честному какой-нибудь `NoFeaturesException`, то его на уровне своей бизнес-логики придется преобразовывать в http-exception (хотя еще правильнее было бы кинуть исключение уровня бизнес-логики, а уже в хэндлере обернуть его)

Voronapxl replied:
Без этих физических объектов невозможен дальнейший расчёт, поэтому выкидываю ошибку как раз

Voronapxl replied:
Теперь возвращаю None и в бизнес-логике проверяю на None. Если None - raise exception

kanootoko commented on app/common/caching/caching_service.py:
```diff
@@ -104,8 +104,17 @@ def parse_task_id(self, task_id: str):
-        return method, scenario_id, phash
\ No newline at end of file
+
+        if len(tail) == 8 and all(c in "0123456789abcdef" for c in tail.lower()):
```
Как насчет того, чтобы заменить вторую часть условия на `tail.lower().strip("0123456789abcdef") == ""`?
Читается попроще (субъективно!), производительность такая же должна быть (мы говорим о строках длиной 8 символов...)

Voronapxl replied:
Тут неправильно будет использовать strip - он удаляет указанные символы только с начала и конца, а не проверяет всю строку. ```tail.lower().strip("0123456789abcdef") == ""   # вернёт True```. Т.е. строка с запрещённым символом внутри пройдёт проверку получается

kanootoko replied:
Подойдёт проверку `== "" `? Попробуй убедить меня ещё раз

kanootoko commented on app/effects_api/effects_controller.py:
```diff
@@ -7,6 +7,7 @@
 
 from app.common.auth.auth import verify_token
 
+from ..common.exceptions.http_exception_wrapper import http_exception
```
А вот тут, видимо, совсем лишний импорт

kanootoko commented on app/effects_api/effects_service.py:
```diff
@@ -58,6 +59,8 @@
     SocioEconomicSchema,
 )
 
+config.change_logger_lvl("DEBUG")
```
Вероятно, случайно осталось в коммите? Иначе не очень понятно, почему именно в этом файле прямо при импорте меняется уровень логгирования

Voronapxl replied:
Да, я забыл убрать

kanootoko commented on app/effects_api/effects_service.py:
```diff
@@ -924,12 +955,12 @@ async def territory_transformation_scenario_after(
-            params.context_func_source_year,
-            token,
-        )
+        # context_blocks, _ = await self.aggregate_blocks_layer_context(
+        #     params.scenario_id,
+        #     params.context_func_zone_source,
+        #     params.context_func_source_year,
+        #     token,
+        # )
```
Код за комментариями

kanootoko commented on app/effects_api/effects_service.py:
```diff
@@ -975,7 +1006,7 @@ async def territory_transformation_scenario_after(
             facade=facade,
             weights=services_weights,
-            max_evals=30,
+            max_evals=25,
```
Магическое число 30 здесь и ниже поменялось на не менее магическое 25. А была бы одна константа/параметр - поменялось бы в одном месте, и было бы более понятно, почему оно такое

Voronapxl replied:
На самом деле это величины для двух разных вещей. max_runs - это число итераций, а max_evals Это сколько итерация внутри max_runs

kanootoko replied:
Ну это ещё один поинт в сторону именованных констант, да

kanootoko commented on app/effects_api/modules/service_type_service.py:
```diff
@@ -75,7 +75,9 @@ async def get_services_with_ids_from_layer(
-    def map_services(names: List[str]):
+    VALID_SERVICE_NAMES = set(SERVICE_TYPES_MAPPING.values())
+
+    def map_services(names: List[str]) -> List[dict]:
```
Вот эта функция и та, что ниже, создаются каждый раз при вызове метода, и при этом не захватывают переменные из локального скоупа. Можно вынести их немного наружу, чтобы сделать микро-оптимизацию.

К слову, не очень понятно, почему `VALID_SERVICE_NAMES` в верхнем регистре, когда это уже обычная переменная в методе

kanootoko commented on app/effects_api/modules/task_service.py:
```diff
@@ -31,18 +31,20 @@ def __init__(
         self.param_hash = params_hash
 
-        self.task_id = f"{method}_{scenario_id}_{self.param_hash}"
+        # self.task_id = f"{method}_{scenario_id}_{self.param_hash}"
```
Что-то лишнее осталось

kanootoko commented on app/effects_api/tasks_controller.py:
```diff
@@ -108,18 +113,42 @@ async def get_territory_transformation_layer(scenario_id: int, service_name: str
+            }
+        )
+
+    elif fc_before and not fc_after:
+        return JSONResponse(
+            content={"before": fc_before, "provision_total_before": provision_before}
+        )
+
+    elif fc_after and not fc_before:
```
Можно ограничиться if'ами без elif'ов, так как во всех ветках return делается

Voronapxl committed:
fix(exception_handlers):
1. New business-level logic exception classes added
2. Deleted comments from code

Voronapxl committed:
fix(exceptions):
1. Fixed exception for physical object requests

Voronapxl committed:
fix(constants):
1. Added max runs and evaluations into constants.py
